[New Tutorial] Solving Problems You Can't See: Using AWS X-Ray and CloudWatch for user-level observability in your serverless microservices applications (#603)
build-on-aws/content: tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md

![AWS X-Ray segments timeline exceptions view](images/x-ray-segments-timeline-exceptions.png)

Before you go fix this issue, take a closer look at the logs provided by AWS X-Ray.  One thing to note here is that the logs displayed in X-Ray don’t contain the exception as you would expect it to.  You will notice in the screenshot above that the log statements that are included in X-Ray include the X-Ray trace id.  AWS X-Ray is using [CloudWatch Logs Insights](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html) to retrieve the logs for the request.  Click on the **View in CloudWatch Logs Insights**.  You should see a query similar to this:

```
fields @log, @timestamp, @message
| filter @message like "05bf2f42-cd3b-4e4b-8876-f97b022525be" or @requestId = "c2eac3ec-d724-496c-9eb0-a112dc2c459d" or @message like "1-6470a77f-746c4b830dc79084524259da"
| sort @timestamp, @message desc
```

> You can see that the query is looking for either the API Gateway request id, AWS Lambda Function request id, or the AWS X-Ray trace id in the message.   When you log messages with the **aws_lambda_powertools** logger and you have also instrumented your function to use AWS X-Ray, it will automatically enrich the log messages with the AWS X-Ray trace id.  However, the exception that was thrown was not logged using the logger and therefore didn’t include the X-Ray trace id.  You can confirm this by proceeding to the logs for the **aws-serverless-shopping-cart-shop-ListCartFunction-\<hash>** AWS Lambda function and viewing the exception message in the logs:  

Reviewer: ```suggestion
You can see that the query is looking for either the API Gateway request id, AWS Lambda Function request id, or the AWS X-Ray trace ID in the message.   When you log messages with the **aws_lambda_powertools** logger and you have also instrumented your function to use AWS X-Ray, it will automatically enrich the log messages with the AWS X-Ray trace ID.  However, the exception that was thrown was not logged using the logger and therefore didn’t include the X-Ray trace id.  You can confirm this by proceeding to the logs for the **aws-serverless-shopping-cart-shop-ListCartFunction-\<hash>** AWS Lambda function and viewing the exception message in the logs:  
```

![AWS Lambda exception log correlation](images/aws-lambda-exception-log-correlation.png)

Because the exception abruptly ended your AWS Lambda Function execution, you weren’t able to end your **annotations** subsegment with your annotations and therefore your trace didn’t include annotations.  You will also notice that upstream requests such as your call to DynamoDB didn’t show up in your trace as well.

Next, you will fix the error and also improve exception handling so that your X-Ray traces aren’t negatively impacted by exceptions.  Proceed to the **backend/shopping-cart-service/list_cart.py** file.  The exception is occuring because of this code block:

```python
    for product in product_list:
        product["sk"] = product["sk"].replace("product#", "")
        if product["sk"] == "8c843a54-27d7-477c-81b3-c21db12ed1c9":
            product.remove()
    xray_recorder.end_subsegment()
```

> The exception was caused by incorrectly using the `remove()` python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:

Reviewer: ```suggestion
The exception was caused by incorrectly using the `remove()` python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:
```
```suggestion
The exception was caused by incorrectly using the `remove()` Python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:
```

> ---

Reviewer: Hello! This is an excellent blog! Great job with the intro. You set up the scene nicely and it invites the reader. The structure is solid too. You have a friendly voice and conversational tone in the piece that plays well with the aphorisms. 

For my edits I added contractions here and there to make it more conversational. I tightened some sentences to eliminate overuse of "that", so it reads more fluidly. 

Since the links are embedded into the copy I removed the brackets ONLY when the links are part of a complete sentence. It makes the blog easier to read digitally. I also capitalized titles and proper nouns. 

Overall, well done!

Author: Thank you for all your edits.  The only concern I have is with making the title too general because then its harder for the reader to understand what they will learn.  Also I believe there were 2 images that you found that you said were unavailable?  I am not sure what the issue is there, these are showing up on my version.

Reviewer: See my notes below on the title and descriptor. 

I still can't see all the images. No idea why. But I trust your end's technology more than mine.

> title: Solving Problems You Can't See - Using AWS X-Ray and CloudWatch for user-level observability in your serverless microservices applications

Reviewer: ```suggestion
title: Solve Problems You Can't See- How AWS X-Ray and CloudWatch Help You Save Stranded Users
```

Author: Hi, I think its important that we include "observability" and "serverless microservices applications" as these statements provide specificity for the tutorial.   We could shorten it to:
Solving Problems You Can't See- How AWS X-Ray and CloudWatch for can provide observability in your serverless microservices applications
 
or something similar

Reviewer: Got it. So, please rewrite the descriptor. The title and the descriptor don't need to repeat each other. 

Keep the title in AP Style: "Solving Problems You Can’t See- How AWS X-Ray and CloudWatch Provide Observability in Serverless Microservices Applications".

Then, rewrite the descriptor without the words "observability" and "serverless microservices applications". It should also be written at 60 characters. 

Something like: 
EX: Avoid blind spots by tracking users on apps built on AWS.

> description: Learn how to use AWS X-Ray to trace and observe your user interactions in a serverless microservices application built on AWS.

Reviewer: ```suggestion
description: Observe users in serverless microservices applications built on AWS.
```

| ------------------- |-------------------------------------------------------------------------------------------------------------------------|
| ✅ AWS Level        | Advanced - 300                                                                                                          |
| ⏱ Time to complete  | 45 minutes                                                                                                              |
| 💰 Cost to complete | Free when using the AWS Free Tier                                                                         |
| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/)<br>- [Git client](https://git-scm.com/downloads/guis) |
| 💻 Code Sample         | [aws-serverless-shopping-cart](https://github.com/aws-samples/aws-serverless-shopping-cart)                             |
| 📢 Feedback            | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?       |
| ⏰ Last Updated     | 2023-09-02                                                                                                              |

> | ToC |

Author: Hi, not sure @scribwell ?
